### PR TITLE
Update Basic Cluster to align with CHIP spec

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem

The currently implemented Basic Cluster is pulled unmodified from
the ZCL spec and does not align with the CHIP defined Basic Cluster.

This PR is the concatenation of #5019 and some updates to support generating writable string attributes.

#### Summary of Changes

- update `general.xml` for Basic cluster
- rename a field in `ta.xml` and `application-basic-cluster.xml` to avoid conflicts
- update `.zap` files
- Update the ZAP templates to support generating writable string attributes
- regenerate artifacts

issue #2877 Update Basic Cluster to align with CHIP spec

### CODE REVIEW

**Comment by vivien-apple** [comment]
For the record, most of the commits of this PR has already been reviewed in #5019. Only the last commit differs from it.

File: examples/chip-tool/templates/commands.zapt [hunk 1]
```diff
@@ -209,7 +217,11 @@ public:
     Write{{asCamelCased parent.name false}}{{asCamelCased name false}}(): ModelCommand("write")
     {
         AddArgument("attr-name", "{{asDelimitedCommand (asCamelCased name)}}");
+        {{#if (isString type)}}
```
Comment (bzbarsky-apple) [inline examples/chip-tool/templates/commands.zapt:220]: Do we need to differentiate here between byte string and not?  Or have we not diverged the storage for those yet?

File: src/app/zap-templates/templates/chip/encoder-src.zapt [hunk 2]
```diff
@@ -193,15 +193,29 @@ PacketBufferHandle encode{{asCamelCased parent.name false}}ClusterRead{{asCamelC
 }
 
 {{#if (isWritableAttribute)}}
-PacketBufferHandle encode{{asCamelCased parent.name false}}ClusterWrite{{asCamelCased name false}}Attribute(uint8_t seqNum, EndpointId destinationEndpoint, {{asUnderlyingZclType type}} {{asCamelCased name}})
+PacketBufferHandle encode{{asCamelCased parent.name false}}ClusterWrite{{asCamelCased name false}}Attribute(uint8_t seqNum, EndpointId destinationEndpoint, {{chipType}} {{asCamelCased name}})
 {
     COMMAND_HEADER("Write{{asCamelCased parent.name false}}{{asCamelCased name false}}", {{parent.define}}_ID);
+    {{#if (isString type)}}
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/chip/encoder-src.zapt:199]: This definitely looks like it needs a bytestring case, like the one command encoding has.

File: src/app/zap-templates/templates/chip/encoder-src.zapt [hunk 3]
```diff
@@ -193,15 +193,29 @@ PacketBufferHandle encode{{asCamelCased parent.name false}}ClusterRead{{asCamelC
...
+        return PacketBufferHandle();
+    }
+
+    {{/if}}
     buf.Put8(kFrameControlGlobalCommand)
        .Put8(seqNum)
        .Put8(ZCL_WRITE_ATTRIBUTES_COMMAND_ID)
        .Put16({{asHex attributeCode 4}})
        .Put8({{atomicTypeId}})
+    {{#if (isString type)}}
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/chip/encoder-src.zapt:213]: And this needs a bytestring case as well.

File: src/app/zap-templates/templates/chip/helper.js [hunk 4]
```diff
@@ -248,6 +248,13 @@ function getAttributes(pkgId, options)
       // Enhanced the attribute with 'atomidId', 'discrete', chipType properties for convenience.
       att.atomicTypeId = atomic.atomicId;
       att.discrete     = atomic.discrete;
+
+      if (StringHelper.isString(att.type)) {
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/chip/helper.js:252]: What does that do for byte strings?

File: examples/chip-tool/templates/commands.zapt (hunk 1 above)
Comment (vivien-apple) [reply examples/chip-tool/templates/commands.zapt:226]: I have updated the PR to handle `OCTET_STRING` and `CHAR_STRING` more gracefully in general.
I end up using `chip::ByteSpan` for both `OCTET_STRING` and `CHAR_STRING` in the API. High level callers/consumers can implement more pleasant types.

For example the objc API is exposing `OCTET_STRING` as `NSData *`, while `CHAR_STRING` are exposed as `NSString *`.

For this case and for `chip-tool` the input type is `char *`.

File: src/app/zap-templates/templates/chip/helper.js (hunk 4 above)
Comment (vivien-apple) [reply src/app/zap-templates/templates/chip/helper.js:248]: Now, this is `chip::ByteSpan` for everything.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Comment by woody-apple** [comment]
@vivien-apple sorry more conflicts :(

**Comment by vivien-apple** [comment]
Sigh. A new conflict just in-between my rebase :(

**Pull request merged**